realm/realm-library/src/main/java/io/realm/internal/Context.java

```diff
@@ -16,32 +16,60 @@
 
 package io.realm.internal;
 
-import java.lang.ref.Reference;
 import java.lang.ref.ReferenceQueue;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
 
 public class Context {
 
+    // Pool to hold the phantom references.
+    // The size of array for storing phantom references will never decrease. Instead, we use another array to hold the
+    // index of the free slot. When adding the reference, pick the last index from freeIndexList and put the reference
+    // to the corresponding slot.
+    private static class ReferencesPool {
+        ArrayList<NativeObjectReference> arrayList = new ArrayList<>();
+        ArrayList<Integer> freeIndexList = new ArrayList<Integer>();
```

Comment:
According to this, you would be spending 4x the amount of memory: https://speakerdeck.com/romainguy/android-memories?slide=36. You are probably right it will not provide any measurable speed difference, but increasing the number of objects will have a negative impact on the GC, which means that cost might be outside the benchmark.

But benchmarking with ArrayDeque sounds fine to me.